SETI-68: Make logger more tolerant to invalid log levels

gurkanoluc opened the pull request:
Closes [SETI-68](https://deliveroo.atlassian.net/browse/SETI-68)

Basically prints info message incase log level is not supported and falls back to `INFO` level for the non-supported levels.

gurkanoluc committed:
SETI-68: Make logger more tolerant to invalid log levels

gurkanoluc committed:
SETI-68: Replace string concat with replacement

mezis commented on config/bootstrap.rb:
```diff
@@ -32,3 +32,10 @@ def perform
 end
+
+require 'routemaster/mixins/log'
+include Routemaster::Mixins::Log
```
You're adding to the global namespace — not ideal?

Instead of forcing this in the bootstrap, you could for instance choose the path of lazyness and `warn` on the first log call.

gurkanoluc replied:
Hmm that was the way chosen in `bin/worker` so I wanted to stick with that by thinking we wouldn't want everything to be global space of course but some things such as logging helpers might be useful. 

For lazy path, I guess we need to either monkey patch or extend `Logger` to sneak our check in? Extending sounds cleaner to me. What do you think?

mezis replied:
I'm not sure why you'd need to do either patching or extending? Actually I'm not sure why you want to add anything to the bootstrap?
That's why I wrote this:

> Instead of forcing this in the bootstrap, you could for instance choose the path of lazyness and `warn` on the first log call.

gurkanoluc replied:
Ah, sorry I thought you meant lazily checking that inside `warn`, `info` calls. You meant checking that inside `_log` method. Yeah, that makes sense.

mezis commented on routemaster/mixins/log.rb:
```diff
@@ -66,6 +70,22 @@ def _smart_backtrace(e)
+        if _log_level_invalid?
+          Logger::INFO
+        else
+          Logger.const_get(_fetch_log_level)
```
Since this is only called once — you could call this, rescue the `NameError` exception and log there. Something like this:

```ruby
def _log_level
  level_name = ENV.fetch('ROUTEMASTER_LOG_LEVEL', 'INFO')
  Logger.const_get level_name
rescue NameError
  warn "log level #{level_name} is invalid, defaulting to INFO"
  return Logger::INFO 
end
```

gurkanoluc replied:
`NameError` seemed too generic to catch to me as it might be raised in any case you misspelled the variable name or something that kind, so it we catch, wouldn't it make debugging bit harder for this part of the code?

mezis commented on routemaster/mixins/log.rb:
```diff
@@ -66,6 +70,22 @@ def _smart_backtrace(e)
+        ENV.fetch('ROUTEMASTER_LOG_LEVEL', 'INFO')
+      end
+
+      def _log_levels
```
You probably don't want to hardcode these — the set of valid values is exactly
```ruby
Logger::Severity.constants.map(&:to_s)
```

gurkanoluc committed:
SETI-68: Remove logging from bootstrap and lazily warn

gurkanoluc committed:
SETI-68: Fix broken tests of Log mixin

SETI-68: Another try to fix log mixin tests

SETI-68: Fix the flaky test finally

gurkanoluc committed:
SETI-68: Fix broken tests of Log mixin

gurkanoluc commented:
@mezis I had to update the tests for `Mixin::Log` to get them working properly. Could you have another look please? :) Thanks.

mezis commented on spec/mixins/log_spec.rb:
```diff
@@ -0,0 +1,64 @@
+
+        ENV['ROUTEMASTER_LOG_LEVEL'] = 'FOO'
+        allow(Logger).to receive(:new).and_return(logger)
+      end
```
This block is duplicated from above. Also, it's quite odd to stub the `Logger` here, as it's more or less what we're testing?

mezis replied:
(also, almost none of the Routemaster codebase uses stubs or doubles — on purpose, as this is critical code and we want to avoid flakyness)

gurkanoluc committed:
SETI-68: Refactor logger and extract it to its own class

gurkanoluc committed:
SETI-68: Fix broken specs

tompave-roo commented on spec/mixins/log_spec.rb:
```diff
@@ -0,0 +1,5 @@
+require 'spec_helper'
+require 'routemaster/mixins/log'
+
+describe Routemaster::Mixins::Log do
```
Is this meant to contain any test?

gurkanoluc replied:
Nope. In the first implementation I changed the logic that was happening inside `_log` method then later we decided that for something that's there just to add syntactic sugar it had become too complex and we extracted big part of its own class and cleaned the content of this file. I think it's good to keep the file still in case we add something.

tompave-roo commented on routemaster/services/logger.rb:
```diff
@@ -0,0 +1,74 @@
+require 'logger'
```
if you're requiring `logger` then, to be consistent, you should also require 'delegate'

tompave-roo commented on routemaster/mixins/log.rb:
```diff
@@ -8,15 +8,7 @@ module Log
-            logger.formatter = method(:_formatter)
-          end
-        end
+        @@_logger ||= Services::Logger.new
```
I know that this was already here in the previous version, but is this thread safe?
I guess it doesn't matter because it's not holding state, and the object is always configured in the same way.

gurkanoluc replied:
Yep, it's not thread safe. But It's going to be initialised only once before the threads start their lives and also as you said since it's not holding any state no prob if it's initialised twice.

tompave-roo commented:
A couple of comments, but the changes make sense.

gurkanoluc committed:
SETI-68: require delegate explicitly
